File: windows/QTube.Windows/Services/YtdlpService.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using QTube.Windows.Models;

namespace QTube.Windows.Services
{
    public class DownloadRequest
    {
        public string Url { get; set; } = string.Empty;
        public string OutputDirectory { get; set; } = string.Empty;
        public DownloadMode Mode { get; set; } = DownloadMode.Video;
        public VideoQuality Quality { get; set; } = VideoQuality.Best;
        public AudioFormat AudioFormat { get; set; } = AudioFormat.Mp3;
        public bool DownloadSubtitles { get; set; } = false;
        public string? CustomFilename { get; set; }
        public BrowserCookieSource BrowserCookieSource { get; set; } = BrowserCookieSource.Auto;
    }

    public class ProgressUpdate
    {
        public double Fraction { get; set; }
        public long DownloadedBytes { get; set; }
        public long? TotalBytes { get; set; }
        public double? SpeedBytesPerSecond { get; set; }
        public int? EtaSeconds { get; set; }
    }

    public class YtdlpService
    {
        private Process? _process;
        private readonly object _lock = new();
        private bool _isCancelled;

        public event Action<string>? TitleReceived;
        public event Action<ProgressUpdate>? ProgressReceived;
        public event Action? ProcessingStarted;
        public event Action<string>? FilePathReceived;

        public async Task DownloadAsync(DownloadRequest request, CancellationToken cancellationToken = default)
        {
            lock (_lock)
            {
                _isCancelled = false;
            }

            string? ytDlpPath = ToolLocator.YtDlp();
            if (string.IsNullOrEmpty(ytDlpPath))
            {
                throw new FileNotFoundException("Không tìm thấy yt-dlp.exe. Vui lòng kiểm tra thư mục Tools.");
            }

            string? ffmpegPath = ToolLocator.Ffmpeg();
            if (string.IsNullOrEmpty(ffmpegPath))
            {
                throw new FileNotFoundException("Không tìm thấy ffmpeg.exe. Vui lòng kiểm tra thư mục Tools.");
            }

            string? quickJsPath = ToolLocator.QuickJs();
            string? denoPath = ToolLocator.Deno();

            await ExecuteAttemptAsync(request, ytDlpPath, ffmpegPath, quickJsPath, denoPath, attempt: 1, cancellationToken);
        }

        private async Task ExecuteAttemptAsync(
            DownloadRequest request,
            string ytDlpPath,
            string ffmpegPath,
            string? quickJsPath,
            string? denoPath,
            int attempt,
            CancellationToken cancellationToken)
        {
            var args = BuildArguments(request, ffmpegPath, quickJsPath, denoPath);

            var psi = new ProcessStartInfo
            {
                FileName = ytDlpPath,
                UseShellExecute = false,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                CreateNoWindow = true,
                StandardOutputEncoding = Encoding.UTF8,
                StandardErrorEncoding = Encoding.UTF8
            };

            foreach (var arg in args)
            {
                psi.ArgumentList.Add(arg);
            }

            var process = new Process { StartInfo = psi };

            lock (_lock)
            {
                if (_isCancelled) throw new OperationCanceledException();
                _process = process;
            }

            var diagnosticLines = new List<string>();

            try
            {
                process.Start();

                var registration = cancellationToken.Register(() =>
                {
                    Cancel();
                });

                var stdoutTask = Task.Run(async () =>
                {
                    using var reader = process.StandardOutput;
                    string? line;
                    while ((line = await reader.ReadLineAsync()) != null)
                    {
                        ParseLine(line);
                        if (!string.IsNullOrWhiteSpace(line))
                        {
                            diagnosticLines.Add(line);
                            if (diagnosticLines.Count > 20) diagnosticLines.RemoveAt(0);
                        }
                    }
                });

                var stderrTask = Task.Run(async () =>
                {
                    using var reader = process.StandardError;
                    string? line;
                    while ((line = await reader.ReadLineAsync()) != null)
                    {
                        if (!string.IsNullOrWhiteSpace(line))
                        {
                            diagnosticLines.Add(line);
                            if (diagnosticLines.Count > 20) diagnosticLines.RemoveAt(0);
                        }
                    }
                });

                await Task.WhenAll(stdoutTask, stderrTask, process.WaitForExitAsync(cancellationToken));
                registration.Dispose();

                lock (_lock)
                {
                    if (_isCancelled || cancellationToken.IsCancellationRequested)
                    {
                        throw new OperationCanceledException();
                    }
                }

                if (process.ExitCode != 0)
                {
                    string diagnostics = string.Join("\n", diagnosticLines).ToLowerInvariant();

                    bool isBotOrRateLimit = diagnostics.Contains("sign in to confirm you're not a bot")
                        || diagnostics.Contains("sign in to confirm you’re not a bot")
                        || diagnostics.Contains("http error 429")
                        || diagnostics.Contains("http error 403");

                    if (attempt < 2 && isBotOrRateLimit)
                    {
                        await Task.Delay(1500, cancellationToken);
                        await ExecuteAttemptAsync(request, ytDlpPath, ffmpegPath, quickJsPath, denoPath, attempt + 1, cancellationToken);
                        return;
                    }

                    if (diagnostics.Contains("sign in to confirm you're not a bot"))
                    {
                        throw new InvalidOperationException("YouTube yêu cầu xác minh tài khoản. Vui lòng chọn trình duyệt (Edge/Chrome) đã đăng nhập.");
                    }

                    throw new InvalidOperationException($"Quá trình tải thất bại (Mã lỗi {process.ExitCode}): {string.Join(" ", diagnosticLines)}");
                }
            }
            finally
            {
                lock (_lock)
                {
                    _process = null;
                }
                process.Dispose();
            }
        }

        public void Cancel()
        {
            lock (_lock)
            {
                _isCancelled = true;
                if (_process != null && !_process.HasExited)
                {
                    try
                    {
                        _process.Kill(entireProcessTree: true);
                    }
                    catch
                    {
                        // Process may have already terminated
                    }
                }
            }
        }

        private void ParseLine(string line)
        {
            if (string.IsNullOrWhiteSpace(line)) return;

            if (line.StartsWith("[BT_TITLE]"))
            {
                string title = line.Substring("[BT_TITLE]".Length).Trim();
                TitleReceived?.Invoke(title);
            }
            else if (line.StartsWith("[BT_FILE]"))
            {
                string path = line.Substring("[BT_FILE]".Length).Trim();
                FilePathReceived?.Invoke(path);
            }
            else if (line.StartsWith("[BT_PROGRESS]"))
            {
                string json = line.Substring("[BT_PROGRESS]".Length).Trim();
                try
                {
                    using var doc = JsonDocument.Parse(json);
                    var root = doc.RootElement;

                    var update = new ProgressUpdate();

                    if (root.TryGetProperty("downloaded_bytes", out var dlElem) && dlElem.TryGetInt64(out var dl))
                    {
                        update.DownloadedBytes = dl;
                    }

                    if (root.TryGetProperty("total_bytes", out var totalElem) && totalElem.TryGetInt64(out var total))
                    {
                        update.TotalBytes = total;
                    }
                    else if (root.TryGetProperty("total_bytes_estimate", out var estElem) && estElem.TryGetInt64(out var est))
                    {
                        update.TotalBytes = est;
                    }

                    if (update.TotalBytes > 0)
                    {
                        update.Fraction = Math.Clamp((double)update.DownloadedBytes / update.TotalBytes.Value, 0.0, 1.0);
                    }

                    if (root.TryGetProperty("speed", out var speedElem) && speedElem.TryGetDouble(out var speed))
                    {
                        update.SpeedBytesPerSecond = speed;
                    }

                    if (root.TryGetProperty("eta", out var etaElem) && etaElem.TryGetInt32(out var eta))
                    {
                        update.EtaSeconds = eta;
                    }

                    ProgressReceived?.Invoke(update);
                }
                catch
                {
                    // Ignore JSON parse errors for incomplete chunks
                }
            }
            else if (line.Contains("[ExtractAudio]") || line.Contains("[Merger]") || line.Contains("[FixupM3u8]"))
            {
                ProcessingStarted?.Invoke();
            }
        }

        private List<string> BuildArguments(
            DownloadRequest request,
            string ffmpegPath,
            string? quickJsPath,
            string? denoPath)
        {
            string ffmpegDir = Path.GetDirectoryName(ffmpegPath) ?? AppDomain.CurrentDomain.BaseDirectory;

            var args = new List<string>
            {
                "--ignore-config",
                "--no-plugin-dirs",
                "--newline",
                "--no-color",
                "--continue",
                "--no-overwrites",
                "--progress",
                "--progress-delta", "0.5",
                "--progress-template", "download:[BT_PROGRESS]%(progress)j",
                "--print", "before_dl:[BT_TITLE]%(title)s",
                "--print", "after_move:[BT_FILE]%(filepath)s",
                "--ffmpeg-location", ffmpegDir,
                "--paths", request.OutputDirectory,
                "--extractor-args", "youtube:player_client=ios,mweb,web",
                "--socket-timeout", "30",
                "--retries", "10",
                "--fragment-retries", "10",
                "--retry-sleep", "fragment:exp=1:20",
                "--sleep-requests", "1.0"
            };

            // Custom or default filename output template
            if (!string.IsNullOrWhiteSpace(request.CustomFilename))
            {
                string safeName = string.Join("-", request.CustomFilename.Split(Path.GetInvalidFileNameChars())).Trim();
                args.Add("--output");
                args.Add($"{safeName} [%(id)s].%(ext)s");
            }
            else
            {
                args.Add("--output");
                args.Add("%(title).180B [%(id)s].%(ext)s");
            }

            // JavaScript engine to evade YouTube bot challenge
            if (!string.IsNullOrEmpty(quickJsPath))
            {
                args.Add("--js-runtimes");
                args.Add($"quickjs:{quickJsPath}");
            }
            else if (!string.IsNullOrEmpty(denoPath))
            {
                args.Add("--js-runtimes");
                args.Add($"deno:{denoPath}");
            }

            // Browser Cookies
            string? cookieBrowser = request.BrowserCookieSource.GetYtDlpIdentifier();
            if (!string.IsNullOrEmpty(cookieBrowser))
            {
                args.Add("--cookies-from-browser");
                args.Add(cookieBrowser);
            }
            else
            {
                args.Add("--user-agent");
                args.Add("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/128.0.0.0 Safari/537.36");
            }

            // Mode & Format selection
            if (request.Mode == DownloadMode.Audio)
            {
                string ext = request.AudioFormat.GetFileExtension();
                args.Add("--extract-audio");
                args.Add("--audio-format");
                args.Add(ext);
                args.Add("--audio-quality");
                args.Add("0");
                args.Add("--embed-thumbnail");
                args.Add("--add-metadata");
            }
            else
            {
                int? maxHeight = request.Quality.GetMaximumHeight();
                string format;
                if (maxHeight.HasValue)
                {
                    format = $"bestvideo[height<={maxHeight.Value}][ext=mp4]+bestaudio[ext=m4a]/bestvideo[height<={maxHeight.Value}]+bestaudio/best[height<={maxHeight.Value}]/best";
                }
                else
                {
                    format = "bestvideo[ext=mp4]+bestaudio[ext=m4a]/bestvideo+bestaudio/best";
                }

                args.Add("-f");
                args.Add(format);
                args.Add("--merge-output-format");
                args.Add("mp4");
                args.Add("--embed-thumbnail");
                args.Add("--add-metadata");
            }

            if (request.DownloadSubtitles)
            {
                args.Add("--write-subs");
                args.Add("--sub-langs");
                args.Add("all");
                args.Add("--embed-subs");
            }

            // Target URL
            args.Add(request.Url);

            return args;
        }
    }
}

File: Sources/QTube/Services/YTDLPService.swift
import Foundation

struct DownloadRequest {
    let url: URL
    let outputDirectory: URL
    let mode: DownloadMode
    let quality: VideoQuality
    let audioFormat: AudioFormat
    let downloadSubtitles: Bool
    let customFilename: String?
    let browserCookieSource: BrowserCookieSource

    init(
        url: URL,
        outputDirectory: URL,
        mode: DownloadMode,
        quality: VideoQuality,
        audioFormat: AudioFormat = .mp3,
        downloadSubtitles: Bool = false,
        customFilename: String? = nil,
        browserCookieSource: BrowserCookieSource
    ) {
        self.url = url
        self.outputDirectory = outputDirectory
        self.mode = mode
        self.quality = quality
        self.audioFormat = audioFormat
        self.downloadSubtitles = downloadSubtitles
        self.customFilename = customFilename
        self.browserCookieSource = browserCookieSource
    }
}

enum DownloadEvent {
    case title(String)
    case expectedTotalBytes(Int64)
    case progress(DownloadProgressSnapshot)
    case processing
    case outputFile(String)
}

enum DownloadServiceError: LocalizedError {
    case missingTool(String)
    case launchFailed(String)
    case commandFailed(String)
    case youtubeVerificationRequired
    case browserCookiesUnavailable(String)

    var errorDescription: String? {
        switch self {
        case .missingTool(let name): return "Không tìm thấy \(name) trong ứng dụng."
        case .launchFailed(let message): return "Không thể bắt đầu tải: \(message)"
        case .commandFailed(let message): return message
        case .youtubeVerificationRequired:
            return "YouTube yêu cầu xác minh. Hãy chọn trình duyệt đã đăng nhập rồi thử lại."
        case .browserCookiesUnavailable(let browser):
            return "Không thể đọc cookie từ \(browser). Hãy mở trình duyệt, đăng nhập YouTube hoặc chọn trình duyệt khác."
        }
    }

    var requiresBrowserCookies: Bool {
        switch self {
        case .youtubeVerificationRequired, .browserCookiesUnavailable:
            return true
        default:
            return false
        }
    }
}

final class YTDLPService {
    private let stateLock = NSLock()
    private var process: Process?
    private var isCancelled = false
    private let progressAccumulator = ProgressAccumulator()

    func start(
        request: DownloadRequest,
        onEvent: @escaping (DownloadEvent) -> Void,
        completion: @escaping (Result<Void, Error>) -> Void
    ) {
        stateLock.lock()
        isCancelled = false
        stateLock.unlock()

        guard let ytDLP = ToolLocator.ytDLP() else {
            completion(.failure(DownloadServiceError.missingTool("yt-dlp")))
            return
        }
        guard let ffmpeg = ToolLocator.ffmpeg() else {
            completion(.failure(DownloadServiceError.missingTool("ffmpeg")))
            return
        }

        execute(
            request: request,
            attempt: 1,
            ytDLP: ytDLP,
            ffmpeg: ffmpeg,
            onEvent: onEvent,
            completion: completion
        )
    }

    private func checkCancelled() -> Bool {
        stateLock.lock()
        defer { stateLock.unlock() }
        return isCancelled
    }

    private func execute(
        request: DownloadRequest,
        attempt: Int,
        ytDLP: URL,
        ffmpeg: URL,
        onEvent: @escaping (DownloadEvent) -> Void,
        completion: @escaping (Result<Void, Error>) -> Void
    ) {
        if checkCancelled() {
            completion(.failure(CancellationError()))
            return
        }

        let clientStrategy: String
        switch attempt {
        case 1:
            clientStrategy = "ios,mweb,web"
        case 2:
            clientStrategy = "mweb,tv_embedded,ios"
        default:
            clientStrategy = "tv_embedded,mweb,web"
        }

        let task = Process()
        let outputPipe = Pipe()
        task.executableURL = ytDLP
        task.arguments = arguments(
            for: request,
            ffmpeg: ffmpeg,
            deno: ToolLocator.deno(),
            quickjs: ToolLocator.quickjs(),
            playerClients: clientStrategy
        )
        task.standardOutput = outputPipe
        task.standardError = outputPipe
        var environment = ProcessInfo.processInfo.environment
        environment["LC_ALL"] = "C"
        environment["LANG"] = "C"
        task.environment = environment

        stateLock.lock()
        if isCancelled {
            stateLock.unlock()
            completion(.failure(CancellationError()))
            return
        }
        process = task
        stateLock.unlock()

        do {
            try task.run()
        } catch {
            stateLock.lock()
            process = nil
            stateLock.unlock()
            completion(.failure(DownloadServiceError.launchFailed(error.localizedDescription)))
            return
        }

        DispatchQueue.global(qos: .utility).async { [weak self] in
            guard let self else { return }
            var pending = ""
            var diagnosticLines: [String] = []
            let handle = outputPipe.fileHandleForReading

            while true {
                let data = handle.readData(ofLength: 4096)
                if data.isEmpty { break }
                pending += String(decoding: data, as: UTF8.self)

                if let lastNewline = pending.lastIndex(of: "\n") {
                    let completeChunk = pending[..<lastNewline]
                    pending = String(pending[pending.index(after: lastNewline)...])

                    completeChunk.enumerateLines { line, _ in
                        self.parse(line: line, onEvent: onEvent)
                        if !line.isEmpty {
                            diagnosticLines.append(line)
                            if diagnosticLines.count > 12 {
                                diagnosticLines.removeFirst(diagnosticLines.count - 12)
                            }
                        }
                    }
                }
            }

            if !pending.isEmpty {
                self.parse(line: pending, onEvent: onEvent)
                diagnosticLines.append(pending)
            }

            task.waitUntilExit()
            self.stateLock.lock()
            self.process = nil
            let cancelled = self.isCancelled
            self.stateLock.unlock()

            if cancelled || task.terminationReason == .uncaughtSignal || task.terminationStatus == 15 {
                completion(.failure(CancellationError()))
                return
            }

            if task.terminationStatus == 0 {
                completion(.success(()))
                return
            }

            let diagnostics = diagnosticLines.joined(separator: "\n")
            let normalizedDiagnostics = diagnostics.lowercased()

            let isBotOrVerificationError = normalizedDiagnostics.contains("sign in to confirm you're not a bot")
                || normalizedDiagnostics.contains("sign in to confirm you’re not a bot")
                || normalizedDiagnostics.contains("unable to extract video data")
                || normalizedDiagnostics.contains("http error 429")
                || normalizedDiagnostics.contains("http error 403")

            // Auto retry with alternative client if attempt < 2
            if attempt < 2 && isBotOrVerificationError && !self.checkCancelled() {
                DispatchQueue.global(qos: .utility).asyncAfter(deadline: .now() + 1.5) { [weak self] in
                    guard let self, !self.checkCancelled() else {
                        completion(.failure(CancellationError()))
                        return
                    }
                    self.execute(
                        request: request,
                        attempt: attempt + 1,
                        ytDLP: ytDLP,
                        ffmpeg: ffmpeg,
                        onEvent: onEvent,
                        completion: completion
                    )
                }
                return
            }

            if normalizedDiagnostics.contains("sign in to confirm you're not a bot")
                || normalizedDiagnostics.contains("sign in to confirm you’re not a bot") {
                completion(.failure(DownloadServiceError.youtubeVerificationRequired))
                return
            }
            if normalizedDiagnostics.contains("cookie"),
               normalizedDiagnostics.contains("permission denied")
                || normalizedDiagnostics.contains("could not find")
                || normalizedDiagnostics.contains("failed to decrypt") {
                completion(.failure(
                    DownloadServiceError.browserCookiesUnavailable(request.browserCookieSource.displayName)
                ))
                return
            }
            let useful = diagnosticLines
                .reversed()
                .first(where: { $0.localizedCaseInsensitiveContains("error") })
                ?? diagnosticLines.suffix(3).joined(separator: " ")
            let message = useful.isEmpty
                ? "Tải thất bại (mã lỗi \(task.terminationStatus))."
                : useful
            completion(.failure(DownloadServiceError.commandFailed(message)))
        }
    }

    func cancel() {
        stateLock.lock()
        isCancelled = true
        let runningProcess = process
        stateLock.unlock()
        guard let runningProcess, runningProcess.isRunning else { return }
        runningProcess.terminate()
    }

    func arguments(
        for request: DownloadRequest,
        ffmpeg: URL,
        deno: URL? = nil,
        quickjs: URL? = nil,
        playerClients: String = "ios,mweb,web"
    ) -> [String] {
        var result = [
            "--ignore-config",
            "--no-plugin-dirs",
            "--newline",
            "--no-color",
            "--continue",
            "--no-overwrites",
            "--progress",
            "--progress-delta",
            "0.5",
            "--progress-template",
            "download:[BT_PROGRESS]%(progress)j",
            "--print",
            "before_dl:[BT_TITLE]%(title)s",
            "--print",
            "before_dl:[BT_META]%(requested_formats)j",
            "--print",
            "after_move:[BT_FILE]%(filepath)s",
            "--ffmpeg-location",
            ffmpeg.deletingLastPathComponent().path,
            "--paths",
            request.outputDirectory.path,
            "--output",
            {
                if let custom = request.customFilename, !custom.isEmpty {
                    let sanitized = custom
                        .replacingOccurrences(of: "/", with: "-")
                        .replacingOccurrences(of: ":", with: "-")
                        .replacingOccurrences(of: "\"", with: "")
                        .trimmingCharacters(in: .whitespacesAndNewlines)
                    return "\(sanitized) [%(id)s].%(ext)s"
                } else {
                    return "%(title).180B [%(id)s].%(ext)s"
                }
            }(),
            "--extractor-args",
            "youtube:player_client=\(playerClients)",
            "--socket-timeout",
            "30",
            "--retries",
            "10",
            "--fragment-retries",
            "10",
            "--retry-sleep",
            "fragment:exp=1:20",
            "--sleep-requests",
            "1.0"
        ]

        if let quickjs {
            result += ["--js-runtimes", "quickjs:\(quickjs.path)"]
        } else if let deno {
            result += ["--js-runtimes", "deno:\(deno.path)"]
        }

        if let browser = request.browserCookieSource.ytDLPIdentifier {
            result += ["--cookies-from-browser", browser]
        } else {
            result += [
                "--user-agent",
                "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.5 Safari/605.1.15"
            ]
        }

        switch request.mode {
        case .audio:
            let ext = request.audioFormat.fileExtension
            result += [
                "--extract-audio",
                "--audio-format", ext,
                "--audio-quality", "0",
                "--embed-thumbnail",
                "--add-metadata"
            ]
        case .video:
            if let height = request.quality.maximumHeight {
                result += [
                    "--format",
                    "bv*[height<=\(height)]+ba/b[height<=\(height)]",
                    "--merge-output-format",
                    "mp4"
                ]
            } else {
                result += ["--format", "bv*+ba/b", "--merge-output-format", "mp4"]
            }

            if request.downloadSubtitles {
                result += [
                    "--write-sub",
                    "--write-auto-sub",
                    "--sub-lang", "vi,en",
                    "--convert-subs", "srt"
                ]
            }
        }

        result.append(request.url.absoluteString)
        return result
    }

    private func parse(line rawLine: String, onEvent: (DownloadEvent) -> Void) {
        let line = rawLine.trimmingCharacters(in: .whitespacesAndNewlines)
        if line.hasPrefix("[BT_TITLE]") {
            onEvent(.title(String(line.dropFirst("[BT_TITLE]".count))))
            return
        }
        if line.hasPrefix("[BT_FILE]") {
            onEvent(.outputFile(String(line.dropFirst("[BT_FILE]".count))))
            return
        }
        if line.hasPrefix("[BT_META]") {
            let json = String(line.dropFirst("[BT_META]".count))
            if let total = progressAccumulator.consumeMetadataJSON(json) {
                onEvent(.expectedTotalBytes(total))
            }
            return
        }
        if line.hasPrefix("[BT_PROGRESS]") {
            let json = String(line.dropFirst("[BT_PROGRESS]".count))
            if let snapshot = progressAccumulator.consumeProgressJSON(json) {
                onEvent(.progress(snapshot))
            }
            return
        }
        if line.contains("[Merger]") || line.contains("[ExtractAudio]") || line.contains("[VideoConvertor]") {
            onEvent(.processing)
        }
    }
}
